libsolidity/analysis/TypeChecker.cpp

```diff
@@ -2078,6 +2086,112 @@ void TypeChecker::typeCheckABIEncodeFunctions(
 	}
 }
 
+void TypeChecker::typeCheckABIEncodeCallFunction(FunctionCall const& _functionCall)
+{
+	vector<ASTPointer<Expression const>> const& arguments = _functionCall.arguments();
+
+	// Expecting first argument to be the function pointer and second to be a tuple.
+	if (arguments.size() != 2)
+	{
+		m_errorReporter.typeError(
+			6219_error,
+			_functionCall.location(),
+			"Expected two arguments: a function pointer followed by a tuple."
+		);
+		return;
+	}
+
+	auto const functionPointerType = dynamic_cast<FunctionTypePointer>(type(*arguments.front()));
+
+	if (!functionPointerType)
+	{
+		m_errorReporter.typeError(
+			5511_error,
+			arguments.front()->location(),
+			"Expected first argument to be a function pointer, not \"" +
+			type(*arguments.front())->canonicalName() +
+			"\"."
+		);
+		return;
+	}
+
+	if (functionPointerType->kind() != FunctionType::Kind::External)
+	{
+		SecondarySourceLocation ssl{};
+
+		if (functionPointerType->hasDeclaration())
+			ssl.append("Function is declared here:", functionPointerType->declaration().location());
+		m_errorReporter.typeError(
+			3509_error,
+			arguments[0]->location(),
+			ssl,
+			"Function must be \"public\" or \"external\". Did you forget to precede \"this.\"?"
+		);
+	}
+
+	// There are no checks we can do if all types are accepted
+	if (functionPointerType->takesArbitraryParameters())
```

Comment:
Do you have an example here this is true after the check above?

Reply:
Yes, in one of the test cases:
```
	function fail6() public returns(bytes memory) {
		return abi.encodeCall(abi.encodeCall, (1, 2, 3, "test"));
	}
// TypeError 3509: (619-633): Function must be "public" or "external". Did you forget to precede "this."?
// TypeError 6716: (619-633): Function takes arbitrary parameters. No type-checking possible.
```

However, that's only possible because I don't bail out of the function in the check above this one..

Reply:
Yeah, I think it is better to have a proper assertion here, this looks rather dangerous.